### PR TITLE
Script editor: Rewrite multiline move logic

### LINKED ISSUE #68731: Moving a multi-line selection in the script editor using alt+arrow down has weird wrapping behaviour at the bottom
### Godot version

4.0 beta 4

### System information

Windows 11

### Issue description

Moving a multi-line selection using alt+arrow down causes the code to wrap around itself when you've reached the bottom. The bottom line seems to be deselected, after which the rest of the seleciton moves underneath it, which repeats until only a single line is selected. This doesn't happen at the top of the script editor and it doesn't happen in Godot 3.5.1.

### Steps to reproduce

1. Create a script with multiple lines of code (the default code will do) and open it in the script editor.
2. Select more than one line of code in that script.
3. Using alt+arrow down, hold the down arrow until the selected code stops moving completely.
4. Observe that the order of the previously selected code block has changed, and that the selected code now only is the bottom line of the script rather than the previously selected block of code.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by RumblingTurtle**
Fix proposal for #68731. Preserves all carets and their selections intact on alt+up/down shortcut

*Bugsquad edit:*
- Fixes #68731

### CODE REVIEW

File: editor/code_editor.cpp [hunk 1]
```diff
@@ -1290,90 +1290,92 @@ void CodeTextEditor::convert_case(CaseStyle p_case) {
...
+
+			if (text_editor->has_selection(c)) {
+				int from_line = text_editor->get_selection_from_line(c);
+				int from_col = text_editor->get_selection_from_column(c);
+				int to_line = text_editor->get_selection_to_line(c);
+				int to_column = text_editor->get_selection_to_column(c);
+				caret_group_parameters.push_back(Vector<int>{ from_line, from_col, to_line, to_column, cursor_line, cursor_column });
+			} else {
+				caret_group_parameters.push_back(Vector<int>{ -1, -1, -1, -1, cursor_line, cursor_column });
 			}
```
Comment (Paulb23) [inline editor/code_editor.cpp:1358]: `Nit`
```suggestion
			if (!text_editor->has_selection(c)) {
				caret_group_parameters.push_back(Vector<int>{ -1, -1, -1, -1, cursor_line, cursor_column });
				continue;
			}
			int from_line = text_editor->get_selection_from_line(c);
			int from_col = text_editor->get_selection_from_column(c);
			int to_line = text_editor->get_selection_to_line(c);
				int to_column = text_editor->get_selection_to_column(c);
			caret_group_parameters.push_back(Vector<int>{ from_line, from_col, to_line, to_column, cursor_line, cursor_column });
```

File: editor/code_editor.cpp [hunk 2]
```diff
@@ -1290,90 +1290,92 @@ void CodeTextEditor::convert_case(CaseStyle p_case) {
...
 
-					carets_to_remove.push_back(c);
-					i = j - 1;
-					swaped_caret = true;
-					break;
-				}
+		Vector<int> new_group{ c };
+		std::pair<int, int> group_border;
+		if (text_editor->has_selection(c))
+			group_border = std::make_pair<int, int>(text_editor->get_selection_from_line(c), text_editor->get_selection_to_line(c));
```
Comment (Paulb23) [inline editor/code_editor.cpp:1306]: Always use braces please.

File: editor/code_editor.cpp [hunk 3]
```diff
@@ -1290,90 +1290,92 @@ void CodeTextEditor::convert_case(CaseStyle p_case) {
...
+			i = j;
+			if (next_end_pos == current_start_pos || next_end_pos + 1 == current_start_pos) {
+				group_border = std::pair<int, int>(next_start_pos, std::get<1>(group_border));
+				new_group.push_back(c_next);
+				// If the last caret is added to the current group there is no need to process it again
+				if (j + 1 == caret_edit_order.size() - 1)
+					i++;
+			} else {
 				break;
 			}
```
Comment (Paulb23) [inline editor/code_editor.cpp:1328]: `Nit`
```suggestion
			if (next_end_pos != current_start_pos && next_end_pos + 1 != current_start_pos) {
				break;
			}
			group_border = std::pair<int, int>(next_start_pos, std::get<1>(group_border));
			new_group.push_back(c_next);
			// If the last caret is added to the current group there is no need to process it again
			if (j + 1 == caret_edit_order.size() - 1) {
				i++;
			}
```

**Review by Paulb23 — DISMISSED** [review]
This looks great, couple small style changes commented inline.

Can we use `#include "core/templates/pair.h"` instead of `std::pair`.

File: editor/code_editor.cpp (hunk 2 above)
Comment (RumblingTurtle) [reply editor/code_editor.cpp]: Sorry, I was pointed to the same mistake int a different PR but forgot to add them here

**Comment by RumblingTurtle** [comment]
> This looks great, couple small style changes commented inline.
> 
> Can we use `#include "core/templates/pair.h"` instead of `std::pair`.

Sorry, missed the pair part. I'll fix it

**Comment by Paulb23** [comment]
LGTM, any chance you could squash the commits.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**